File: backend/seed.py
"""Seed script to create initial admin user and initial data"""

import random
import sys
from datetime import date

from database import SessionLocal
from repositories import (
    CategoryRepository,
    IncomeTypeRepository,
    MonthRepository,
    PeriodRepository,
    UserRepository,
)
from services.month_service import MonthService
from utils.auth import get_password_hash


def generate_random_color() -> str:
    """
    Generate a random color in hex format
    Generates a truly random hex color with good visibility
    Based on frontend logic: minimum value of 50 ensures colors aren't too dark, max is 255
    """
    # Generate random RGB values
    # Using a range that ensures good visibility (avoiding very dark colors)
    # Minimum value of 50 ensures colors aren't too dark, max is 255
    r = random.randint(50, 255)
    g = random.randint(50, 255)
    b = random.randint(50, 255)

    # Convert to hex format
    def to_hex(value: int) -> str:
        hex_str = hex(value)[2:]  # Remove '0x' prefix
        return hex_str.zfill(2)  # Pad with zero if needed

    return f"#{to_hex(r)}{to_hex(g)}{to_hex(b)}"


def seed_admin_user():
    """Create initial admin user if it doesn't exist"""
    db = SessionLocal()
    try:
        user_repository = UserRepository(db)

        # Check if admin user already exists
        existing_user = user_repository.get_by_email("admin@email.com")
        if existing_user:
            # Update existing user to ensure they are admin
            if not existing_user.is_admin:
                user_repository.update(existing_user, {"is_admin": True})
                print("✓ Admin user updated to have admin privileges!")
            else:
                print("Admin user already exists. Skipping seed.")
            return

        # Create admin user
        admin_user = user_repository.create(
            {
                "email": "admin@email.com",
                "hashed_password": get_password_hash("admin"),
                "full_name": "Administrator",
                "is_active": True,
                "is_admin": True,
            }
        )

        print("✓ Admin user created successfully!")
        print("  Email: admin@email.com")
        print("  Password: admin")
        print(f"  User ID: {admin_user.id}")

    except Exception as e:
        print(f"Error creating admin user: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        db.close()


def seed_current_month():
    """Create current month if it doesn't exist"""
    db = SessionLocal()
    try:
        month_repository = MonthRepository(db)
        month_service = MonthService(month_repository)

        today = date.today()
        year = today.year
        month = today.month

        # Check if current month already exists
        existing_month = month_repository.get_by_year_month(year, month)
        if existing_month:
            print(
                f"✓ Current month ({month_service._generate_month_name(year, month)}) already exists. Skipping."
            )
            return

        # Create current month
        month_data = {"year": year, "month": month}
        month_service.create(month_data, "system")
        print(
            f"✓ Current month ({month_service._generate_month_name(year, month)}) created successfully!"
        )

    except Exception as e:
        print(f"Error creating current month: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        db.close()


def seed_categories():
    """Create initial categories if they don't exist"""
    db = SessionLocal()
    try:
        category_repository = CategoryRepository(db)

        categories = [
            "Groceries",
            "Transportation",
            "Insurance",
            "Subscriptions",
            "Rent/Utilities",
            "Going Out",
            "Purchases",
            "Health",
            "Investment/Savings",
            "Withdraw",
        ]

        created_count = 0
        skipped_count = 0

        for category_name in categories:
            existing_category = category_repository.get_by_name(category_name)
            if existing_category:
                skipped_count += 1
                continue

            color = generate_random_color()
            category_repository.create(category_name, color=color, user_name="system")
            created_count += 1

        if created_count > 0:
            print(f"✓ Created {created_count} categories")
        if skipped_count > 0:
            print(f"  {skipped_count} categories already exist")

    except Exception as e:
        print(f"Error creating categories: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        db.close()


def seed_income_types():
    """Create initial income types if they don't exist"""
    db = SessionLocal()
    try:
        income_type_repository = IncomeTypeRepository(db)

        income_types = [
            "Salary",
            "Carry Over",
            "Side Hustle",
            "Tax Return",
            "Investment Return",
            "Bonus",
        ]

        created_count = 0
        skipped_count = 0

        for income_type_name in income_types:
            existing_income_type = income_type_repository.get_by_name(income_type_name)
            if existing_income_type:
                skipped_count += 1
                continue

            color = generate_random_color()
            income_type_repository.create(income_type_name, color=color, user_name="system")
            created_count += 1

        if created_count > 0:
            print(f"✓ Created {created_count} income types")
        if skipped_count > 0:
            print(f"  {skipped_count} income types already exist")

    except Exception as e:
        print(f"Error creating income types: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        db.close()


def seed_periods():
    """Create initial periods if they don't exist"""
    db = SessionLocal()
    try:
        period_repository = PeriodRepository(db)

        periods = [
            "On Demand",
            "1st Period",
            "2nd Period",
        ]

        created_count = 0
        skipped_count = 0

        for period_name in periods:
            existing_period = period_repository.get_by_name(period_name)
            if existing_period:
                skipped_count += 1
                continue

            color = generate_random_color()
            period_repository.create(period_name, color=color, user_name="system")
            created_count += 1

        if created_count > 0:
            print(f"✓ Created {created_count} periods")
        if skipped_count > 0:
            print(f"  {skipped_count} periods already exist")

    except Exception as e:
        print(f"Error creating periods: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        db.close()


def seed_all():
    """Seed all initial data"""
    print("Seeding initial data...")
    print()
    seed_admin_user()
    print()
    seed_current_month()
    print()
    seed_categories()
    print()
    seed_income_types()
    print()
    seed_periods()
    print()
    print("✓ All initial data seeded successfully!")


if __name__ == "__main__":
    seed_all()

File: backend/src/seed.ts
/**
 * Seed script — creates initial admin user and default data.
 *
 * Usage:
 *   bun run src/seed.ts
 *
 * Idempotent via the seed_records table: a row with
 * seed_id = 'initial_data_v1' marks the initial seed as complete and
 * subsequent runs become a no-op.
 */

import { eq } from 'drizzle-orm';
import { db, sqlite } from './db/connection';
import {
  categories,
  incomeTypes,
  months,
  periods,
  seedRecords,
  users,
} from './db/schema';
import { hashPassword } from './utils/auth';

const SEED_INITIAL_DATA = 'initial_data_v1';

const ADMIN_EMAIL = 'admin@email.com';
const ADMIN_PASSWORD = 'admin1';
const ADMIN_NAME = 'Administrator';

const DEFAULT_CATEGORIES = [
  'Groceries',
  'Transportation',
  'Insurance',
  'Subscriptions',
  'Rent/Utilities',
  'Going Out',
  'Purchases',
  'Health',
  'Investment/Savings',
  'Withdraw',
];

const DEFAULT_INCOME_TYPES = [
  'Salary',
  'Carry Over',
  'Side Hustle',
  'Tax Return',
  'Investment Return',
  'Bonus',
];

const DEFAULT_PERIODS = ['On Demand', '1st Period', '2nd Period'];

const MONTH_NAMES = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December',
];

function now(): string {
  return new Date().toISOString();
}

/** #rrggbb with each channel in [50, 255] for good visibility. */
function randomColor(): string {
  const channel = () => Math.floor(Math.random() * (255 - 50 + 1)) + 50;
  const hex = (v: number) => v.toString(16).padStart(2, '0');
  return `#${hex(channel())}${hex(channel())}${hex(channel())}`;
}

async function hasSeedRun(seedId: string): Promise<boolean> {
  const [row] = await db
    .select()
    .from(seedRecords)
    .where(eq(seedRecords.seed_id, seedId))
    .limit(1);
  return !!row;
}

async function markSeedComplete(seedId: string): Promise<void> {
  await db.insert(seedRecords).values({ seed_id: seedId, executed_at: now() });
}

async function seedAdminUser(): Promise<void> {
  const [existing] = await db
    .select()
    .from(users)
    .where(eq(users.email, ADMIN_EMAIL))
    .limit(1);

  if (existing) {
    if (!existing.is_admin) {
      await db.update(users).set({ is_admin: true, updated_at: now() }).where(eq(users.id, existing.id));
      console.log('✓ Existing admin user updated to have admin privileges');
    } else {
      console.log('  Admin user already exists — skipping');
    }
    return;
  }

  const hashed = await hashPassword(ADMIN_PASSWORD);
  const timestamp = now();
  await db.insert(users).values({
    email: ADMIN_EMAIL,
    hashed_password: hashed,
    full_name: ADMIN_NAME,
    is_active: true,
    is_admin: true,
    created_at: timestamp,
    updated_at: timestamp,
    created_by: 'system',
    updated_by: 'system',
  });
  console.log('✓ Admin user created');
  console.log(`  Email: ${ADMIN_EMAIL}`);
  console.log(`  Password: ${ADMIN_PASSWORD}`);
}

async function seedCurrentMonth(): Promise<void> {
  const today = new Date();
  const year = today.getFullYear();
  const month = today.getMonth() + 1;
  const name = `${MONTH_NAMES[month - 1]} ${year}`;

  const [existing] = await db
    .select()
    .from(months)
    .where(eq(months.name, name))
    .limit(1);
  if (existing) {
    console.log(`  Current month (${name}) already exists — skipping`);
    return;
  }

  const startDate = `${year}-${String(month).padStart(2, '0')}-01`;
  const lastDay = new Date(year, month, 0).getDate();
  const endDate = `${year}-${String(month).padStart(2, '0')}-${String(lastDay).padStart(2, '0')}`;

  const timestamp = now();
  await db.insert(months).values({
    year,
    month,
    name,
    start_date: startDate,
    end_date: endDate,
    is_closed: false,
    created_at: timestamp,
    updated_at: timestamp,
    created_by: 'system',
    updated_by: 'system',
  });
  console.log(`✓ Current month (${name}) created`);
}

async function seedCategories(): Promise<void> {
  let created = 0;
  let skipped = 0;
  const timestamp = now();

  for (const name of DEFAULT_CATEGORIES) {
    const [existing] = await db
      .select()
      .from(categories)
      .where(eq(categories.name, name))
      .limit(1);
    if (existing) {
      skipped++;
      continue;
    }
    await db.insert(categories).values({
      name,
      color: randomColor(),
      created_at: timestamp,
      updated_at: timestamp,
      created_by: 'system',
      updated_by: 'system',
    });
    created++;
  }

  if (created > 0) console.log(`✓ Created ${created} categories`);
  if (skipped > 0) console.log(`  ${skipped} categories already exist`);
}

async function seedIncomeTypes(): Promise<void> {
  let created = 0;
  let skipped = 0;
  const timestamp = now();

  for (const name of DEFAULT_INCOME_TYPES) {
    const [existing] = await db
      .select()
      .from(incomeTypes)
      .where(eq(incomeTypes.name, name))
      .limit(1);
    if (existing) {
      skipped++;
      continue;
    }
    await db.insert(incomeTypes).values({
      name,
      color: randomColor(),
      created_at: timestamp,
      updated_at: timestamp,
      created_by: 'system',
      updated_by: 'system',
    });
    created++;
  }

  if (created > 0) console.log(`✓ Created ${created} income types`);
  if (skipped > 0) console.log(`  ${skipped} income types already exist`);
}

async function seedPeriods(): Promise<void> {
  let created = 0;
  let skipped = 0;
  const timestamp = now();

  for (const name of DEFAULT_PERIODS) {
    const [existing] = await db
      .select()
      .from(periods)
      .where(eq(periods.name, name))
      .limit(1);
    if (existing) {
      skipped++;
      continue;
    }
    await db.insert(periods).values({
      name,
      color: randomColor(),
      created_at: timestamp,
      updated_at: timestamp,
      created_by: 'system',
      updated_by: 'system',
    });
    created++;
  }

  if (created > 0) console.log(`✓ Created ${created} periods`);
  if (skipped > 0) console.log(`  ${skipped} periods already exist`);
}

async function main(): Promise<void> {
  if (await hasSeedRun(SEED_INITIAL_DATA)) {
    console.log(`Seed '${SEED_INITIAL_DATA}' has already run — skipping`);
    return;
  }

  console.log('Seeding initial data…\n');
  await seedAdminUser();
  console.log();
  await seedCurrentMonth();
  console.log();
  await seedCategories();
  console.log();
  await seedIncomeTypes();
  console.log();
  await seedPeriods();
  console.log();

  await markSeedComplete(SEED_INITIAL_DATA);
  console.log(`✓ Seed '${SEED_INITIAL_DATA}' complete`);
}

main()
  .then(() => {
    sqlite.close();
  })
  .catch((err) => {
    console.error('Seed failed:', err);
    try {
      sqlite.close();
    } catch {
      // ignore
    }
    process.exit(1);
  });
